### PR TITLE
Add Lua function to pick records via name hash

### PR DESCRIPTION
**Pull request by devicenull**
### Short description
This adds a Lua function to return a record based on a weighted hash of the DNS record name.  One use case here is to consistently return the same IP address for a particular cache server based on what subdomain is requesting the data.

This is pretty much a copy and paste of pickwhashed

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: docs/lua-records/functions.rst [hunk 1]
```diff
@@ -226,6 +226,29 @@ Record creation functions
...
+  test.example.com will always resolve to the same IP, but test2.example.com may go elsewhere.
+
+  This works similar to round-robin load balancing, but has the advantage of making traffic
+  for the same domain always end up on the same server which can help cache hit rates.
+
+  This function also works for CNAME or TXT records.
+
+  An example::
+
+    mydomain.example.com    IN    LUA    A ("picknamehashed({                             "
```
Comment (Habbie) [inline docs/lua-records/functions.rst:247]:
```suggestion
    mydomain.example.com    IN    LUA    A ("picknamehashed({                          "
```

File: regression-tests.auth-py/test_LuaRecords.py [hunk 2]
```diff
@@ -790,6 +792,33 @@ def testWHashed(self):
...
+        Basic picknamehashed() test with a set of A records
+        As the name is hashed, we should always get the same IP back for the same record name.
+        """
+
+        queries = [
+            {
+                'expected': dns.rrset.from_text('test.namehashed.example.org.', 0,
+                                       dns.rdataclass.IN, 'A',
+                                       '1.2.3.4'),
+                'query': dns.message.make_query('test.namehashed.example.org', 'A')
```
Comment (Habbie) [inline regression-tests.auth-py/test_LuaRecords.py:807]: can you put `query` above `expected` in both dicts?

File: docs/lua-records/functions.rst (hunk 1 above)
Comment (Habbie) [reply docs/lua-records/functions.rst:247]: also, perhaps the example makes more sense as a wildcard

**Review by Habbie — CHANGES_REQUESTED** [review]
great work! I posted a couple of nits

**Comment by devicenull** [comment]
@Habbie What are your thoughts on if you *want* this or not?  For our uses I was able to reimplement this entirely in Lua, and we'll probably go with that instead.

I'm perfectly willing to just close this MR unless you think this is going to be a useful feature for other people going forward.

I'm also willing to spend the time to get this across the line.  I don't have any real strong preferences either way.

**Comment by Habbie** [comment]
> What are your thoughts on if you _want_ this or not?

_want_ is a very strong word. I think it's a neat feature, it comes with tests, which means it will be easy to maintain/keep, so I'd like to merge this after minor cleanup.

**Comment by devicenull** [comment]
Code updated :)

**Review by Habbie — APPROVED** [review]

**Comment by Habbie** [comment]
rebased to fix a conflict

**Comment by Habbie** [comment]
> rebased to fix a conflict

I did not rebase correctly there. Should be fixed now.

**Pull request merged**